server/routes/results.js

```diff
@@ -153,4 +153,165 @@ router.get("/all", async (req, res) => {
 
+// Helper function to calculate commonsensicality
+// const calculateCommonsensicality = async (statementIds) => {
+//   const result = {};
+//   for (const statementId of statementIds) {
+//     const answersForStatement = await answers.findAll({
+//       where: { statementId },
+//       include: [
+//         {
+//           model: statements,
+//           as: "statement",
+//           attributes: [],
+//         },
+//       ],
+//       attributes: [
+//         "I_agree",
+//         "perceived_commonsense",
+//         [Sequelize.col("statement.statementMedian"), "statementMedian"],
+//       ],
+//       raw: true,
+//     });
+
+//     const total = answersForStatement.length;
+//     if (total === 0) {
+//       result[statementId] = 0;
+//       continue;
+//     }
+
+//     // Calculate consensus (ci)
+//     const consensusSum = answersForStatement.reduce(
+//       (acc, answer) => acc + answer.I_agree,
+//       0
+//     );
+//     const consensus = 2 * Math.abs((consensusSum / total) - 0.5);
+
+//     // Calculate majority
+//     const majority = consensusSum / total >= 0.5 ? 1 : 0;
+
+//     // Calculate awareness (ai)
+//     const awareness = answersForStatement.reduce(
+//       (acc, answer) => acc + (answer.perceived_commonsense === majority ? 1 : 0),
+//       0
+//     ) / total;
+
+//     // Calculate commonsensicality (mi)
+//     result[statementId] = Math.sqrt(consensus * awareness);
+//   }
+//   return result;
+// };
+
+// //endpoint to get commonsensicality scores for each question
+// router.post(
+//   "/commonsensicality",
+//   [body("statementIds").isArray().withMessage("statementIds must be an array")],
+//   async (req, res) => {
+//     const errors = validationResult(req);
+//     if (!errors.isEmpty()) {
+//       return res.status(400).json({ errors: errors.array() });
+//     }
+
+//     const { statementIds } = req.body;
+//     console.log("Received statementIds:", statementIds);
+
+//     try {
+//       const commonsensicalityScores = await calculateCommonsensicality(statementIds);
+//       console.log("Calculated commonsensicalityScores:", commonsensicalityScores);
+//       res.json(commonsensicalityScores);
+//     } catch (error) {
+//       console.error("Error calculating commonsensicality:", error);
+//       res.status(500).json({ error: error.message });
+//     }
+//   }
+// );
+
+// Helper function to calculate agreement percentages
+const calculateAgreementPercentage = async (statementIds) => {
+  const result = {};
+  for (const statementId of statementIds) {
+    const answersForStatementAll = await answers
+    .findAll({
+      include: [
+        {
+          model: statements,
+          as: "statement",
+          attributes: [],
+        },
+      ],
+      attributes: [
+        "sessionId",
+        "id",
+        "statementId",
+        "I_agree",
+        "others_agree",   
+      ],
+      raw: true,
+      where: {
+        statementId: statementId
+      }
+    });
```

Comment:
Instead of looping through all ids, which is very resource intensive use the syntax
```
where: {
       statementId: { $in: [1,2,3,4]  }
}

```

Reply:
the `$in` operator didn't work as it caused my array `answersForStatementAll` to return 0 results, but I changed the SQL query to be```
 where: {
      statementId: statementIds,
    },
```
and changed the looping & filtering logic accordingly so it works now -- should only call the SQL query once so its much more resource efficient